content/asm_7.md

```diff
@@ -1,153 +1,268 @@
+        fprintf(stderr, "Error: this program must have 1 command line argument\n");
+        return EXIT_FAILURE;
+    }
 
-	return 0;
+    printf("The argument length is - %d\n", my_strlen(argv[1]));
+
+    return 0;
 }
 ```
 
-Here we can see the same 2 variables as in previous example and inline assembly definition. First of all we put 1 to rax and rdi registers (write system call number, and stdout) as we did it in our plain assembly hello world. Next we do similar operation with rsi and rdi registers but first operands starts with % symbol instead $. It means str is the output operand referred by %1 and len second output operand referred by %2, so we put values of str and len to rsi and rdi with %n notation, where n is number of output operand. Also there is %% prefixed to the register name.
+This is straightforward C code which I think not needed to be explained in details. It checks that exactly one command-line argument was passed, then calls the `my_strlen` function to determine its length. The result is printed to standard output. As I said - pretty straightforward. The single part which interests us is the implementation of the `my_strlen` function. This function is written in assembly, but not in C.
+
+Let's take a look at the implementation of this function:
 
+```assembly
+;; Definition of the text section
+section .text
+        ;; Reference to the entry point of our program
+        global my_strlen
+
+;; Function that returns the length of the string passed via the first argument
+my_strlen:
+        ;; Reset the register value to zero. It will be returned from the
+        ;; function with the result.
+        xor rax, rax
+.loop:
+        ;; Compare the first element in the given string with the `NUL` terminator (end of the string).
+        cmp byte [rdi + rax], 0
+        ;; If we reached the `NUL` terminator exit from the function.
+        je .done
+        ;; Increase the counter which stores the length of the string.
+        inc rax
+        ;; Repeat the operations above while we didn't reach the end of the string.
+        jmp .loop
+.done:
+        ;; Exit from the function and return the result in the `rax` register.
+        ret
 ```
-    This helps GCC to distinguish between the operands and registers. operands have a single % as prefix
+
+The way function's parameters are passed to our assembly function follows the same calling convention which we have seen in the section above. The first argument, which our case is a pointer to the string, passed in the `rdi` register. At the start of the function we clear the value of the `rax` register, since this will serve as our counter. Each time we encounter a character in the string, the loop increments `rax` using it as an accumulator for the string’s length. The loop continues character by character until it reaches the `NUL` terminator, which signals the end of the string. At that point, the value stored in rax is returned to the caller as the final result.
```

Comment:
```suggestion
This function's parameters are passed to our assembly function by following the same calling convention as in the section above. The first argument, which in our case is a pointer to the string, is passed in the `rdi` register. At the start of the function, we clear the value of the `rax` register, since it will serve as our counter. Each time we encounter a character in the string, the loop increments `rax`, using it as an accumulator for the string’s length. The loop continues character by character until it reaches the `NUL` terminator, which signals the end of the string. At that point, the value stored in `rax` is returned to the caller as the final result.
```